ts/src/test/Exchange/base/test.sharedMethods.ts

```diff
@@ -31,46 +54,65 @@ function assertStructure (exchange, method, entry, format, emptyNotAllowedFor =
+            const emptyAllowedForThisKey = exchange.inArray (i, emptyAllowedFor);
+            const value = entry[i];
+            // check when:
+            // - it's not inside "allowe empty values" list
+            // - it's not undefined
+            if (emptyAllowedForThisKey && (value === undefined)) {
+                continue;
             }
+            assert (value !== undefined, i.toString () + ' index is expected to have a value' + logText);
             // because of other langs, this is needed for arrays
-            assert (assertType (exchange, entry, i, format), i.toString () + ' index does not have an expected type ' + logText);
+            assert (assertType (exchange, skippedProperties, entry, i, format), i.toString () + ' index does not have an expected type ' + logText);
         }
     } else {
         assert (typeof entry === 'object', 'entry is not an object' + logText);
         const keys = Object.keys (format);
         for (let i = 0; i < keys.length; i++) {
             const key = keys[i];
-            const keyStr = key.toString ();
-            assert ((key in entry), keyStr + ' key is missing from structure' + logText);
-            if (exchange.inArray (key, emptyNotAllowedFor)) {
-                // if it was in needed keys, then it should have value.
-                assert (entry[key] !== undefined, key + ' key has an null value, but is expected to have a value' + logText);
+            if (key in skippedProperties) {
+                continue;
+            }
+            assert (key in entry, 'key "' + stringValue (key) + '" is missing from structure' + logText);
+            const emptyAllowedForThisKey = exchange.inArray (key, emptyAllowedFor);
+            const value = entry[key];
+            // check when:
+            // - it's not inside "allowe empty values" list
+            // - it's not undefined
+            if (emptyAllowedForThisKey && (value === undefined)) {
+                continue;
+            }
+            // if it was in needed keys, then it should have value.
+            assert (value !== undefined, 'key "' + stringValue (key) + '" is expected to have a value' + logText);
+            // add exclusion for info key, as it can be any type
+            if (key !== 'info') {
+                assert (assertType (exchange, skippedProperties, entry, key, format), 'key "' + stringValue (key) + '" is neither undefined, neither of expected type' + logText);
             }
-            assert (assertType (exchange, entry, key, format), key + ' key is neither undefined, neither of expected type' + logText);
         }
     }
 }
 
-function assertTimestamp (exchange, method, entry, nowToCheck = undefined, keyName : any = 'timestamp') {
+function assertTimestamp (exchange, skippedProperties, method, entry, nowToCheck = undefined, keyName : any = 'timestamp') {
```

Comment:
skippedProperties is ignored here, is that intended?

Reply:
yes, that is not used there atm

Reply:
@ttodua but why?

Reply:
As I have said, there are many exchange skips now and checks are not much strict. in next updates I will make some check stricter and I will need to have that property there, so I will need `skipProperties` in `asserTimestamp` too. if you look deeper in that method, you will see that I have loyal `if (value != undefined)` which means that atm it does not check the value strictly for exchanges. however, in next updates I am going to remove that `value != undefined` check and directly assume the value to be provided. However, as some exchanges (10-15) will not be able to pass here, I will add those exhanges->methodName->timestamp prop in skips.

Reply:
@ttodua I understand, but with 3 lines we can make it consistent with others, so why not make it?

We just need to add
```
    if (key in skippedProperties) {
        return;
    }
```

Reply:
the thing was that it didnt break any test so I've just ignored that method in this PR, but ok,i'm adding it now.

Reply:
added (note, it accepts also arrays, i.e. ohlcv, so `key` might be index too, not string, so only `x in y` will not be enough, that's why I've added .safeValue).